src/app/clusters/access-control-server/access-control-server.cpp

```diff
@@ -15,4 +15,423 @@
+    CHIP_ERROR ReadExtension(AttributeValueEncoder & aEncoder);
+    CHIP_ERROR WriteAcl(AttributeValueDecoder & aDecoder);
+    CHIP_ERROR WriteExtension(AttributeValueDecoder & aDecoder);
+};
+
+CHIP_ERROR AccessControlAttribute::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    switch (aPath.mAttributeId)
+    {
+    case Clusters::AccessControl::Attributes::Acl::Id:
+        return ReadAcl(aEncoder);
+    case Clusters::AccessControl::Attributes::Extension::Id:
+        return ReadExtension(aEncoder);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AccessControlAttribute::ReadAcl(AttributeValueEncoder & aEncoder)
+{
+    AccessControlEntryCodec codec;
+    AccessControl::EntryIterator iterator;
+
+    GetAccessControl().Entries(iterator);
+
+    CHIP_ERROR err = aEncoder.EncodeList([&](const auto & encoder) -> CHIP_ERROR {
+        while (iterator.Next(codec.entry) == CHIP_NO_ERROR)
+        {
+            encoder.Encode(codec);
+        }
+        return CHIP_NO_ERROR;
+    });
+
+    return err;
+}
+
+CHIP_ERROR AccessControlAttribute::ReadExtension(AttributeValueEncoder & aEncoder)
+{
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AccessControlAttribute::Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder)
+{
+    switch (aPath.mAttributeId)
+    {
+    case Clusters::AccessControl::Attributes::Acl::Id:
+        return WriteAcl(aDecoder);
+    case Clusters::AccessControl::Attributes::Extension::Id:
+        return WriteExtension(aDecoder);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AccessControlAttribute::WriteAcl(AttributeValueDecoder & aDecoder)
+{
+    DataModel::DecodableList<AccessControlEntryCodec> list;
+    ReturnErrorOnFailure(aDecoder.Decode(list));
+
+    size_t oldCount;
+    size_t newCount;
+    size_t maxCount;
+    ReturnErrorOnFailure(GetAccessControl().GetEntryCount(oldCount));
+    ReturnErrorOnFailure(list.ComputeSize(&newCount));
+    ReturnErrorOnFailure(GetAccessControl().GetMaxEntryCount(maxCount));
+    ReturnErrorCodeIf(newCount > maxCount, CHIP_ERROR_INVALID_LIST_LENGTH);
+
+    auto iterator = list.begin();
+    size_t i      = 0;
+    while (iterator.Next())
+    {
+        if (i < oldCount)
+        {
+            ReturnErrorOnFailure(GetAccessControl().UpdateEntry(i, iterator.GetValue().entry));
+        }
+        else
+        {
+            ReturnErrorOnFailure(GetAccessControl().CreateEntry(nullptr, iterator.GetValue().entry));
+        }
+        ++i;
+    }
+    ReturnErrorOnFailure(iterator.GetStatus());
+
+    while (i < oldCount)
+    {
+        --oldCount;
+        ReturnErrorOnFailure(GetAccessControl().DeleteEntry(oldCount));
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AccessControlAttribute::WriteExtension(AttributeValueDecoder & aDecoder)
+{
+    return CHIP_NO_ERROR;
+}
+
+AccessControlAttribute gAttribute;
+
+AccessControl gAccessControl(Examples::GetAccessControlDelegate());
```

Comment:
Both of these should be static or inside an anonymous namespace.  Preferably the latter, along with the various struct definitions above.

Reply:
I agree I'll do that.

Reply:
Please see PR #12690